### PR TITLE
Op_bincount [Numpy]

### PR DESCRIPTION
**Pull request by Tommliu**
## Description ##
Numpy bincount operator

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/numpy/np_bincount_op.cu [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file np_bicount_op-inl.h
```
Comment (xidulu) [inline src/operator/numpy/np_bincount_op.cu]: np_bicount_op-inl.cu

File: python/mxnet/symbol/numpy/_symbol.py [hunk 2]
```diff
@@ -5160,5 +5160,38 @@ def load_json(json_str):
     check_call(_LIB.MXSymbolCreateFromJSON(c_str(json_str), ctypes.byref(handle)))
     return _Symbol(handle)
 
+@set_module('mxnet.symbol.numpy')
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/_symbol.py]: One more blank line above.

File: src/operator/numpy/np_bincount_op-inl.h [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+template<typename xpu>
+void NumpyBincountForwardImpl(const OpContext &ctx,
+                              const NDArray &data,
+                              const NDArray &out,
+                              const size_t &data_n,
+                              const int &minlength);
+
+template<typename xpu>
+
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op-inl.h]: no blank line here.

File: src/operator/numpy/np_bincount_op.cc [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                                   const NDArray &data,
+                                   const NDArray &weights,
+                                   const NDArray &out,
+                                   const size_t &data_n,
+                                   const int &minlength) {
+using namespace mxnet_op;
+BinNumberCount(data, minlength, out, data_n);
+mshadow::Stream<cpu> *s = ctx.get_stream<cpu>();
+MSHADOW_TYPE_SWITCH(data.dtype(), DType, {
+    MSHADOW_TYPE_SWITCH(weights.dtype(), OType, {
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op.cc]: 2-space indentations

File: src/operator/numpy/np_bincount_op.cc [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+template<>
+void NumpyBincountForwardImpl<cpu>(const OpContext &ctx,
+                                   const NDArray &data,
+                                   const NDArray &weights,
+                                   const NDArray &out,
+                                   const size_t &data_n,
+                                   const int &minlength) {
+using namespace mxnet_op;
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op.cc]: 2-space indentations.

File: src/operator/numpy/np_bincount_op.cc [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  });
+}
+
+template<>
+void NumpyBincountForwardImpl<cpu>(const OpContext &ctx,
+                                   const NDArray &data,
+                                   const NDArray &out,
+                                   const size_t &data_n,
+                                   const int &minlength) {
+using namespace mxnet_op;
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op.cc]: 2-space indentations.

File: src/operator/numpy/np_bincount_op.cc [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+      BincountCpu(data.data().dptr<DType>(), out.data().dptr<OType>(), data_n);
+    });
+  });
+}
+
+DMLC_REGISTER_PARAMETER(NumpyBincountParam);
+
+NNVM_REGISTER_OP(_npi_bincount)
+.describe(R"code(
+Count number of occurrences of each value in array of non-negative ints.
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op.cc]: No need for describe here.

File: src/operator/numpy/np_bincount_op.cu [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+ */
+
+#include "./np_bincount_op-inl.h"
+#include "../tensor/util/tensor_util-inl.cuh"
+#include "../tensor/util/tensor_util-inl.h"
+
+namespace mxnet {
+namespace op {
+
+struct BincountFusedKernel{
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op.cu]: `struct BincountFusedKernel {`

File: src/operator/numpy/np_bincount_op.cu [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+template<>
+void NumpyBincountForwardImpl<gpu>(const OpContext &ctx,
+                                   const NDArray &data,
+                                   const NDArray &weights,
+                                   const NDArray &out,
+                                   const size_t &data_n,
+                                   const int &minlength) {
+using namespace mxnet_op;
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op.cu]: 2-space indentations.

File: tests/python/unittest/test_numpy_op.py [hunk 10]
```diff
@@ -5050,6 +5050,56 @@ def hybrid_forward(self, F, a):
...
+            weights = np.array(()).astype(weight_type) if has_weight else None
+        weights_np = weights.asnumpy() if has_weight else None
+        test_bincount = TestBincountWeights(minlength) if has_weight else TestBincount(minlength)
+        if hybridize:
+            test_bincount.hybridize()
+        mx_out = test_bincount(data, weights) if has_weight else test_bincount(data)
+        np_out = _np.bincount(data.asnumpy(), weights_np, minlength)
+        assert mx_out.shape == np_out.shape
+        assert_almost_equal(mx_out.asnumpy(), np_out, rtol=rtol, atol=atol)
+        # No bacward operation for operator bincount at this moment
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: `backward`

File: src/operator/numpy/np_bincount_op-inl.h [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+#include <mxnet/operator_util.h>
+#include <utility>
+#include <vector>
+#include "../mshadow_op.h"
+#include "../mxnet_op.h"
+#include "../operator_common.h"
+#include "../elemwise_op_common.h"
+#include "np_broadcast_reduce_op.h"
+#ifdef __CUDACC__
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op-inl.h]: You can probably put those cuda-only includes in `.cu` file

File: src/operator/numpy/np_bincount_op.cu [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    for (size_t i = 0; i < data_n; i++) {
+      if (h_ptr[i] + 1 > bin) bin = h_ptr[i] + 1;
+    }
+    free(h_ptr);
+    mxnet::TShape s(1, bin);
+    const_cast<NDArray &>(out).Init(s);  // set the output shape forcefully
+  });
+
+  MSHADOW_TYPE_SWITCH(data.dtype(), DType, {
+      MSHADOW_TYPE_SWITCH(out.dtype(), OType, {
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op.cu]: 2-space indentation.

File: src/operator/numpy/np_bincount_op-inl.h [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+              "If minlength is specified, there will be at least this"
+              "number of bins in the output array");
+    DMLC_DECLARE_FIELD(has_weights)
+    .set_default(false)
+    .describe("Determine whether Bincount has weights.");
+  }
+};
+
+inline bool NumpyBincountType(const nnvm::NodeAttrs& attrs,
+                     std::vector<int> *in_attrs,
```
Comment (haojin2) [inline src/operator/numpy/np_bincount_op-inl.h]: alignment:
```c++
inline bool NumpyBincountType(const nnvm::NodeAttrs& attrs,
                              std::vector<int> *in_attrs,
                              std::vector<int> *out_attrs) {
```

**Pull request merged**